ipa-core/src/protocol/context/mod.rs

```diff
@@ -211,6 +211,14 @@ impl ShardedContext for Base<'_, Sharded> {
             .gateway
             .get_shard_receiver(&ChannelId::new(origin, self.gate.clone()))
     }
+
+    fn cross_shard_prss(&self) -> InstrumentedIndexedSharedRandomness<'_> {
```

Comment:
would it make sense to implement this on non-sharded contexts, so that a protocol can use this regardless of sharding (in the non-sharded case, it just acts like normal prss)?

Reply:
I think this is where Rust type system plays nicely for us - in non-sharded world you should never require cross-shard PRSS and use regular PRSS instead. All sharded circuits we write operate over `ShardedContext` that will make this API available. Non-sharded MPC won't compile if it tries to use it, which is what we would want imo

Reply:
got it. so for example with [`gen_prf_key`](https://github.com/private-attribution/ipa/blob/main/ipa-core/src/protocol/ipa_prf/prf_eval.rs#L76) we'd just have two different implementations, one with `Upgradable` constraint on `C`, and another with `Upgradable + Sharded`? (just to make sure I'm understanding correctly.)

Reply:
yea, we will delete non-sharded version anyway, so the only one that we will use is `gen_prf_key` with `C: UpgradableContext + ShardedContext` bound. That will provide access to `cross_shard_prss`